ENIGMAsystem/SHELL/Universal_System/Resources/AssetArray.h

```diff
@@ -0,0 +1,104 @@
+/** Copyright (C) 2019 Robert B. Colton
+***
+*** This file is a part of the ENIGMA Development Environment.
+***
+*** ENIGMA is free software: you can redistribute it and/or modify it under the
+*** terms of the GNU General Public License as published by the Free Software
+*** Foundation, version 3 of the license or any later version.
+***
+*** This application and its source code is distributed AS-IS, WITHOUT ANY
+*** WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+*** FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+*** details.
+***
+*** You should have received a copy of the GNU General Public License along
+*** with this code. If not, see <http://www.gnu.org/licenses/>
+**/
+
+#ifdef INCLUDED_FROM_SHELLMAIN
+#  error This file includes non-ENIGMA STL headers and should not be included from SHELLmain.
+#endif
+
+#ifndef E_ASSET_ARRAY
+#define E_ASSET_ARRAY
+
+#include <vector>
+#include <string>
+#include <utility>
+
+#ifdef DEBUG_MODE
+  #include "Widget_Systems/widgets_mandatory.h" // for show_error
+  #define CHECK_ID(id, ret) \
+    if (!exists(id)) { \
+      enigma_user::show_error("Requested " + asset_type + " asset " + std::to_string(id) + " does not exist.", false); \
+      return ret; \
+    }
+  #define CHECK_ID_V(id) CHECK_ID(id,)
+#else
+  #define CHECK_ID(id, ret)
+  #define CHECK_ID_V(id)
+#endif
+
+namespace enigma {
+
+// Asset storage container designed for dense cache-efficient resource processing.
+template<typename T>
+class AssetArray {
+public:
+  AssetArray(std::string asset_type): asset_type(asset_type) {}
+
+  int add(T asset) {
+    int id = assets.size();
+    assets.emplace_back(true, asset);
+    return id;
+  }
+
+  int assign(int id, T asset) {
+    #ifdef DEBUG_MODE
+    if (id < 0) {
+      enigma_user::show_error("Attempting to assign " + asset_type + " asset " + std::to_string(id) + " to negative index.", false);
+      return id;
+    }
+    #endif
+    if (size_t(id) >= size()) {
+      assets.resize(size_t(id) + 1);
+    }
+    assets[id] = {true,asset};
+    return id;
+  }
+
+  T get(int id) {
+    static T sentinel;
+    CHECK_ID(id,sentinel);
+    return assets[id].second;
+  }
+
+  int replace(int id, T asset) {
+    CHECK_ID(id, -1);
+    assets[id] = {true,asset};
+    return id;
+  }
+
+  int duplicate(int id) {
+    CHECK_ID(id, -1);
+    T asset = assets[id];
+    return add(asset);
+  }
+
+  void destroy(int id) {
+    CHECK_ID_V(id);
+    //TODO: call cleanup method
+    assets[id].first = false;
+  }
+
+  size_t size() const { return assets.size(); }
+  bool exists(int id) { return (id >= 0 && size_t(id) < size() && assets[id].first); }
+
+private:
+  std::vector<std::pair<bool,T>> assets;
+  std::string asset_type;

```

Comment:
This can be a template parameter. You can also make it a static member of Sound, and use, eg, `T::resource_name`.

Reply:
I think I like the idea of template parameter better, but I don't exactly know what you meant. It seems there's multiple ways of doing that but it matters where the string is stored. Can you explain to me how you think it should be passed as a template parameter?

Reply:
You should be able to accept a char[] as a template parameter. If not, you can parameterize on a length and accept a `char[length]`.

Reply:
Looking closer at the template problem, I think I hate doing it that way. So I went and put it in the struct in 225b26cf1659d6a7a253474e0674994e1f110a57 which I'd like to know if you think is better?

Reply:
Yes, that was my other suggestion (although I had said to just make it a constexpr char[] rather than a method to call). Still, this is fine.

Reply:
Ah, I see why I did it that way. Up to this point, many of our resource classes do not have sources because they don't need them. Though, the definition of the asset array that holds them does need a source. Considering that the method can be inlined like the member, and the method can be optimized out like the member in run/compile modes, I don't think there's any other practical difference between the two. I think I'll just keep what I have to avoid scattering the resource classes around.

Worth mentioning that I found an alternative specialization we could do. I was looking to see if we could just grab T's typename directly, but it's only human-readable in MSVC. I don't particularly fancy the specialization to fix GCC.
https://stackoverflow.com/a/4485051

Going to resolve this, because I find what I have the most practical solution.